collectors/cgroups.plugin/sys_fs_cgroup.c

```diff
@@ -857,40 +906,125 @@ static inline void cgroup_read_cpuacct_stat(struct cpuacct_stat *cp) {
+    cp->nr_throttled_delta = cp->nr_throttled >= nr_throttled_last ? cp->nr_throttled - nr_throttled_last : 0;
 
-        unsigned long lines = procfile_lines(ff);
+    cp->updated = 1;
 
-        if(unlikely(lines < 3)) {
-            error("CGROUP: file '%s' should have 3+ lines.", cp->filename);
-            cp->updated = 0;
-            return;
+    if (unlikely(cp->enabled == CONFIG_BOOLEAN_AUTO)) {
+        if (likely(
+                cp->nr_periods || cp->nr_throttled || cp->throttled_time ||
+                netdata_zero_metrics_enabled == CONFIG_BOOLEAN_YES)) {
+            cp->enabled = CONFIG_BOOLEAN_YES;
         }
+    }
+}
 
-        cp->user = str2ull(procfile_lineword(ff, 1, 1));
-        cp->system = str2ull(procfile_lineword(ff, 2, 1));
+static inline void cgroup2_read_cpuacct_stat(struct cpuacct_stat *cp) {
+    static procfile *ff = NULL;
+    if (unlikely(!cp->filename)) {
+        return;
+    }
 
-        cp->updated = 1;
+    ff = procfile_reopen(ff, cp->filename, NULL, PROCFILE_FLAG_DEFAULT);
+    if (unlikely(!ff)) {
+        cp->updated = 0;
+        cgroups_check = 1;
+        return;
+    }
 
-        if(unlikely(cp->enabled == CONFIG_BOOLEAN_AUTO &&
-                    (cp->user || cp->system || netdata_zero_metrics_enabled == CONFIG_BOOLEAN_YES)))
+    ff = procfile_readall(ff);
+    if (unlikely(!ff)) {
+        cp->updated = 0;
+        cgroups_check = 1;
+        return;
+    }
+
+    unsigned long lines = procfile_lines(ff);
+
+    if (unlikely(lines < 3)) {
+        error("CGROUP: file '%s' should have at least 3 lines.", cp->filename);
+        cp->updated = 0;
+        return;
+    }
+
+    unsigned long long nr_periods_last = cp->nr_periods; 
+    unsigned long long nr_throttled_last = cp->nr_throttled; 
+
+    for (unsigned long i = 0; i < lines; i++) {
+        char *s = procfile_lineword(ff, i, 0);
+        uint32_t hash = simple_hash(s);
+
+        if (unlikely(hash == user_usec_hash && !strcmp(s, "user_usec"))) {
+            cp->user = str2ull(procfile_lineword(ff, i, 1));
+        } else if (unlikely(hash == system_usec_hash && !strcmp(s, "system_usec"))) {
+            cp->system = str2ull(procfile_lineword(ff, i, 1));
+        } else if (unlikely(hash == nr_periods_hash && !strcmp(s, "nr_periods"))) {
+            cp->nr_periods = str2ull(procfile_lineword(ff, i, 1));
+        } else if (unlikely(hash == nr_throttled_hash && !strcmp(s, "nr_throttled"))) {
+            cp->nr_throttled = str2ull(procfile_lineword(ff, i, 1));
+        } else if (unlikely(hash == throttled_usec_hash && !strcmp(s, "throttled_usec"))) {
+            cp->throttled_time = str2ull(procfile_lineword(ff, i, 1));
+        }
+    }
+    cp->nr_periods_delta = cp->nr_periods >= nr_periods_last ? cp->nr_periods - nr_periods_last : 0;
+    cp->nr_throttled_delta = cp->nr_throttled >= nr_throttled_last ? cp->nr_throttled - nr_throttled_last : 0;
+
+    cp->updated = 1;
```

Comment:
A simple suggestion, the four lines above are also repeated between the lines `951-954`, so they could be an inline function like this:

static inline void PLEASE_GIVE_ME_A_NAME(struct cpuacct_stat *cp)
{
    cp->nr_periods_delta = cp->nr_periods >= nr_periods_last ? cp->nr_periods - nr_periods_last : 0;
    cp->nr_throttled_delta = cp->nr_throttled >= nr_throttled_last ? cp->nr_throttled - nr_throttled_last : 0;

    cp->updated = 1;
}

Reply:
`cp->updated = 1;` has nothing to do with delta calculation and shouldn't be in one function. I did some modification.